### PR TITLE
destination: add `UriLikeIdentity` and server_name

### PR DESCRIPTION
**Pull request by zaharidichev**
This PR changes the `TlsIdentity` type in the destination API such that: 

- we add an extra `UriLikeIdentity` identity type that should contain identities that are in URI format (e.g. SPIFFE)
- we add a `server_name` to the `TlsIdentity` type. This allows us to differentiate between an SNI value and a TLS Id value. This is mainly needed because in certain identity systems (SPIFFE/SPIRE) the TLS SAN can be in URI form. A URI cannot be used as a SNI extension in a `ClientHello`, so an alternative SNI value needs to be provided. This brings the need to distinguish between these two concepts. 

For context: 
https://github.com/linkerd/linkerd2-proxy/pull/2506

### CODE REVIEW

**Review by mateiidavid — APPROVED** [review]
Looks good to me after playing with this in the prototype. Should we also add a small comment in https://github.com/linkerd/linkerd2-proxy-api/blob/eff6f41ed982460553fd83b331382fc6c99bb769/proto/inbound.proto#L173 to say identity can be either uri or dns?

File: proto/destination.proto [hunk 1]
```diff
@@ -103,12 +103,16 @@ message WeightedAddr {
...
 
-// Which strategy should be used for verifying TLS.
 message TlsIdentity {
   reserved 2;
   reserved "k8s_pod_identity";
 
-  oneof strategy { DnsLikeIdentity dns_like_identity = 1; }
+  oneof strategy {
+    DnsLikeIdentity dns_like_identity = 1;
+    UriLikeIdentity uri_like_identity = 3;
```
Comment (olix0r) [inline proto/destination.proto:112]: If I recall correctly, the name _DnsLikeIdentity_ was chosen to indicate that this isn't actually a DNS name:

1. We don't expect it to be resolvable via DNS.
2. It doesn't actually support all DNS names (i.e., trailing dots are forbidden).

I think in this case, we can just call it UriIdentity.

File: proto/destination.proto [hunk 2]
```diff
@@ -103,12 +103,16 @@ message WeightedAddr {
...
+  }
+
+  ServerName server_name = 4;
```
Comment (olix0r) [inline proto/destination.proto:115]: It seems appropriate to use DnsLikeIdentity for the server_name type to me.  The server name has to be a DNS-like string.

**Review by olix0r — APPROVED** [review]

**Comment by zaharidichev** [comment]
@olix0r renamed `UriIdentity` back to `UriLikeIdentity` as discussed earlier. Are we fine merging that now?

**Comment by zaharidichev** [comment]
Will merge that, we can do a follow up if needed.

**Pull request merged**